### DISCUSSION TITLE
mysql query optimizations

### DISCUSSION THREAD

**Discussion by tonyhouse1982** [photoprism/photoprism#2006] [General]
Hi all, I'm running photoprism on a very limited SBC (odroid n2 with 4gb ram)
running my own mariadb instance and the performances of photoprism UI (expecially latest version) are not so good
While inspecting queries generated by the UI I see queries using a very bad execution plan
example:
`SELECT
  photos.*,
  photos.id AS composite_id,
  files.id AS file_id,
  files.file_uid,
  files.instance_id,
  files.file_primary,
  files.file_sidecar,
  files.file_portrait,
  files.file_video,
  files.file_missing,
  files.file_name,
  files.file_root,
  files.file_hash,
  files.file_codec,
  files.file_type,
  files.file_mime,
  files.file_width,
  files.file_height,
  files.file_aspect_ratio,
  files.file_orientation,
  files.file_main_color,
  files.file_colors,
  files.file_luminance,
  files.file_chroma,
  files.file_projection,
  files.file_diff,
  files.file_duration,
  files.file_size,
  cameras.camera_make,
  cameras.camera_model,
  lenses.lens_make,
  lenses.lens_model,
  places.place_label,
  places.place_city,
  places.place_state,
  places.place_country
FROM
  `photos`
  JOIN files
    ON photos.id = files.photo_id
    AND files.file_missing = 0
    AND files.deleted_at IS NULL
  LEFT JOIN cameras
    ON photos.camera_id = cameras.id
  LEFT JOIN lenses
    ON photos.lens_id = lenses.id
  LEFT JOIN places
    ON photos.place_id = places.id
WHERE (
    files.file_type = 'jpg'
    OR files.file_video = 1
  )
  AND (files.file_error = '')
  AND (photos.deleted_at IS NULL)
  AND (photos.photo_private = 0)
  AND (photos.photo_quality >= 3)
ORDER BY taken_at DESC,
  photos.photo_uid,
  files.file_primary DESC
LIMIT 360 OFFSET 0`

this query (the default one to retrieve images when You login into UI) use a bad query plan, because it uses filesort and doesn't use a proper index to reduce rows examined.
At least valid rows should be indexed and not searched every time... I mean the part regarding files* and photo* conditions....
Is there a plan to focus on database review/performance?

**Comment by graciousgrey** [comment]
Currently we are focusing on multi user support. 
Here's the ticket for performance optimizations https://github.com/photoprism/photoprism/issues/1438 :)

See also: https://docs.photoprism.app/getting-started/troubleshooting/performance/

**Reply by tonyhouse1982** [reply]
thanks I'll read it
